Add new tutorial: resonant circuit tutorial (#480)
precice/tutorials: resonant-circuit/README.md

---
title: Resonant Circuit
keywords: MATLAB
summary: We simulate a two-element LC circuit (one inductor and one capacitor).
---

## Setup

The purpose of this tutorial is to illustrate the usage of preCICE to couple MATLAB code. Two different MATLAB solvers will be coupled to simulate a two-element LC circuit. This type of circuit consists on a very simple system with one inductor and one capacitor:

![LC circuit diagram [1]](images/tutorials-resonant-circuit-diagram.svg)

The circuit is described by the following system of ODEs:

$U(t) = L \frac{\text{d}I}{\text{d}t}$

$I(t) = -C \frac{\text{d}U}{\text{d}t}$

where $I$ is the current and $U$ the voltage of the cirucit.

Each of these equations is going to be solved by a different MATLAB solver. Note that as only one scalar is solved per equation, this is a 0+1 dimensional problem.

> 

Reviewer: You need a new section here to show the config visualization. This is a recent change, see #514. Merge from `develop`, then you can use the tool `toosl/visualize-configs.sh`.

Reviewer 2: Thanks for mentioning it. I merged the branches and updated the README accordingly

---
title: Resonant Circuit
keywords: MATLAB
summary: We simulate a two-element LC circuit (one inductor and one capacitor).
---

## Setup

> The purpose of this tutorial is to illustrate the usage of preCICE to couple MATLAB code. Two different MATLAB solvers will be coupled to simulate a two-element LC circuit. This type of circuit consists on a very simple system with one inductor and one capacitor:

Reviewer: Just a general picky comment (no need to fix now, but keep it in mind for future PRs): we want to keep language a bit more informal and engaging, see [content guidelines](https://precice.org/docs-meta-content-guidelines.html).

Don't be afraid to write exactly as you would imagine explaining something to a colleague over coffee. :smile:

Reviewer 2: Thanks for the tip. I'll keep that in mind :)

![LC circuit diagram [1]](images/tutorials-resonant-circuit-diagram.svg)

The circuit is described by the following system of ODEs:

$U(t) = L \frac{\text{d}I}{\text{d}t}$

$I(t) = -C \frac{\text{d}U}{\text{d}t}$

where $I$ is the current and $U$ the voltage of the cirucit.

Each of these equations is going to be solved by a different MATLAB solver. Note that as only one scalar is solved per equation, this is a 0+1 dimensional problem.

## Configuration

preCICE configuration (image generated using the [precice-config-visualizer](https://precice.org/tooling-config-visualization.html)):

![preCICE configuration visualization](images/tutorials-resonant-circuit-precice-config.png)

## Available solvers

> * <em>MATLAB</em> A solver using the [MATLAB bindings](https://github.com/precice/matlab-bindings).

Reviewer: For consistency and to avoid injecting HTML into Markdown:

```suggestion
* MATLAB. A solver using the [MATLAB bindings](https://precice.org/installation-bindings-matlab.html).
```

>  Before running this tutorial, follow the [instructions](https://github.com/precice/matlab-bindings?tab=readme-ov-file#matlab-bindings) to correctly install the MATLAB bindings.

Reviewer: ```suggestion
 Before running this tutorial, follow the [instructions](https://precice.org/installation-bindings-matlab.html) to correctly install the MATLAB bindings.
```

If that page is not up-to-date, then we need to update it :smile:

Author: True. We need to migrate some content from the `README.md` to the website. But for now, let's the link as suggested above. Currently, the Website provides minimal information and points to the `README.md` for further information. This we can change in the future.

Reviewer: And how will we remember to do that? :smile: 
I would suggest to already use the link to the website and just do a copy-paste of the README now, if possible. Or just open an issue.

## Running the simulation

### MATLAB

For running this example, first get into one of folders with the solvers and open a MATLAB instance.
Afterward, do the same for the second solver.
After adding the MATLAB bindings to the MATLAB path (in both instances), run the following commands:

In the first MATLAB instance one can run the solver for the current:

```MATLAB
coil
```

And in the second MATLAB instance the solver for the voltage:

```MATLAB
capacitor
```

The preCICE configuration file is available as `precice-config.xml`, and it is called directly in the solvers.

#### Running from terminal

> If you prefer to not open the MATLAB GUIs, one can alternatively use two shells instead.

Reviewer: ```suggestion
If you prefer to not open the MATLAB GUIs, you can alternatively use two shells instead.
```

For that, modify the path in the file `matlab-bindings-path.sh` found in the base directory of this tutorial to the path to your MATLAB bindings.

By doing that, you can now open two shells and switch into the directories `capacitor-matlab` and `coil-matlab` and execute the `run.sh` scripts.

> ## Visualization

Reviewer: For consistency:

```suggestion
## Post-processing
```

The solver for the current also records the current and voltage through time and at the end of the simulation saves a plot with the obtained curves, as well as the analytical solution.

> After successfully running the coupling, one can find the curves in the folder `capacitor-matlab` as `Curves.png`.

Reviewer: You could also include this image here (under `resonant-circuit/images/`). In that case, name it `tutorials-resonant-circuit-curves.png`.

## References

> [1] By First Harmonic - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=21991221

Reviewer: Title is missing

Author: ```suggestion
[1] Schematic of a simple parallel LC circuit by First Harmonic - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=21991221
```
